wdl-lint/src/rules/command_shellcheck.rs

```diff
@@ -0,0 +1,357 @@
+                } else {
+                    '_'
+                }
+            })
+            .collect::<String>(),
+    );
+    bash_var
+}
+
+/// retrieve all input and private declations for a task
+fn task_declarations(task: &TaskDefinition) -> HashSet<String> {
+    let mut decls = HashSet::new();
+    if let Some(input) = task.input() {
+        for decl in input.declarations() {
+            decls.insert(decl.name().as_str().to_owned());
+        }
+    }
+
+    for decl in task.declarations() {
+        decls.insert(decl.name().as_str().to_owned());
+    }
+    decls
+}
+
+/// Creates a "ShellCheck lint" diagnostic from a ShellCheckComment
+fn shellcheck_lint(comment: &ShellCheckComment, span: Span) -> Diagnostic {
+    dbg!(
+        Diagnostic::note("shellcheck violations within a command")
+            .with_rule(ID)
+            .with_label(
+                format!("SC{}[{}]: {}", comment.code, comment.level, comment.message),
+                span,
+            )
+            .with_fix("address ShellCheck violations")
+    )
+}
+
+impl Visitor for CommandSectionShellCheck {
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        _: &Document,
+        _: SupportedVersion,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn command_section(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        section: &CommandSection,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        if !program_exists(SHELLCHECK_BIN) {
+            return;
+        }
+
+        // collect declarations so we can ignore placeholder variables
+        let parent_task = section.parent().into_task().expect("parent is a task");
+        let mut decls = task_declarations(&parent_task);
+
+        // replace all placeholders in the command with
+        // dummy bash variables
+        let mut sanitized_command = String::new();
+        if let Some(cmd_parts) = section.strip_whitespace() {
+            cmd_parts.iter().for_each(|part| match part {
+                StrippedCommandPart::Text(text) => {
+                    sanitized_command.push_str(text);
+                }
+                StrippedCommandPart::Placeholder(placeholder) => {
+                    let bash_var = to_bash_var(placeholder);
+                    // we need to save the var so we can suppress later
+                    decls.insert(bash_var.clone());
+                    let mut expansion = String::from("\"$");
+                    expansion.push_str(&bash_var);
+                    expansion.push('"');
+                    sanitized_command.push_str(&expansion);
+                }
+            });
+        } else {
+            // TODO add a diagnostic explaining why this is being skipped?
```

Comment:
This case is that the command used mixed indentation. We may potentially still want to parse and lint it using SC? Or we can throw up our hands and say "you have a weird command block and we aren't SC linting it" 🤷

Reply:
If another diagnostic will catch it and spit out a response, I feel it's best for this rule to silently return. Once the user fixes _that_ diagnostic, then these will start showing up.

If we go this route, ot would also be good to document explicitly above the `if` statement that a `None` result there means there is mixed indentation.

Reply:
Yes, another diagnostic should fire to warn about the mixed indentation. I agree it's fine to silently return here (with a comment explaining this).